File: SecureBlackbox Samples/Office Verifier/java/officeverifier.java
/*
 * SecureBlackbox 2024 Java Edition - Sample Project
 *
 * This sample project demonstrates the usage of SecureBlackbox in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/secureblackbox
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */

import java.io.*;
import secureblackbox.*;

import static secureblackbox.OfficeSignature.*;

public class officeverifier extends ConsoleDemo {
	private static String optval(String[] args, String option) {
		for (int x = 0; x < args.length - 1; x++) {
			if (args[x].equalsIgnoreCase(option)) {
				return args[x + 1];
			}
		}
		return "";
	}

	private static boolean optext(String[] args, String option) {
		for (int x = 0; x < args.length; x++) {
			if (args[x].equalsIgnoreCase(option)) {
				return true;
			}
		}
		return false;
	}

	private static void displayHelp(String errMes) {
		System.out.println(
				"NAME\n" +
				"  officeverifier -- SecureBlackbox OfficeVerifier Demo Application\n\n" +
				"SYNOPSIS\n" +
				"  officeverifier <-input input_file> [-cert certificate_file] [-certpass certificate_password] [-checkrev] [-ignoreerrors] [-offline]\n" +
				"DESCRIPTION\n" +
				"  OfficeVerifier demonstrates the usage of OfficeVerifier from SecureBlackbox.\n" +
				"  Used to verify the signature.\n\n" +
				"  The options are as follows:\n\n" +
				"  -input        An input file to verify (Required).\n\n" +
				"  -cert         The certificate used to sign files.\n\n" +
				"  -certpass     The password for the signing certificate.\n\n" +
				"  -checkrev      Whether certificate revocation information should be checked.\n\n" +
				"  -ignoreerrors  Whether to ignore chain validation errors.\n\n" +
				"  -offline       Whether offline mode be used.\n\n" +
				"EXAMPLES\n" +
				"  officeverifier -input C:\\myfile.doc -output C:\\sign.doc -offline\n\n" +
				"  officeverifier -input C:\\myfile.doc -output C:\\sign.doc -checkrev -ignoreerrors\n"
		);

		if (errMes.length() > 0) {
			System.out.println("Error: " + errMes);
			System.out.println();
		}

		confirmExit();
	}

	private static void confirmExit() {
		System.out.println("Press Enter to exit the demo.");
		input();
	}

	private static String translateSigType(int value) {
		switch (value) {
			case ostBinaryCryptoAPI: return "BinaryCryptoAPI";
			case ostBinaryXML: return "BinaryXML";
			case ostOpenXML: return "OpenXML";
			case ostOpenXPS: return "OpenXPS";
			case ostOpenDocument: return "OpenOffice";
			default: return "Unknown";
		}
	}

	private static String translateDocSig(boolean value) {
		return value ? "Document content is signed" : "Document content is partially signed";
	}

	private static String translateCore(boolean value) {
		return value ? "Document properties are signed" : "Document properties are not signed";
	}

	private static String translateOrigSig(boolean value) {
		return value ? "Signature origin is signed" : "Signature origin is not signed";
	}

	private static String translateSigValidity(int value) {
		switch (value) {
			case svtValid: return "Valid";
			case svtCorrupted: return "Corrupted";
			case svtSignerNotFound: return "Signer not found";
			case svtFailure: return "Failure";
			default: return "Unknown";
		}
	}

	private static String translateChainValidity(int value) {
		switch (value) {
			case cvtValid: return "Valid";
			case cvtValidButUntrusted: return "ValidButUntrusted";
			case cvtInvalid: return "Invalid";
			case cvtCantBeEstablished: return "CantBeEstablished";
			default: return "Unknown";
		}
	}

	public static void main(String[] args) {
		if (args.length == 0) {
			displayHelp("");
			return;
		}

		try {
			OfficeVerifier verifier = new OfficeVerifier();

			if (optext(args, "-input")) {
				verifier.setInputFile(optval(args, "-input"));
			} else {
				displayHelp("-input is required.");
				return;
			}

			if (optext(args, "-cert")) {
				CertificateManager cm = new CertificateManager();
				cm.importFromFile(optval(args, "-cert"), optval(args, "-certpass"));
				verifier.getKnownCertificates().add(cm.getCertificate());
			}

			// Additional options
			if (optext(args, "-checkrev")) {
				verifier.setRevocationCheck(OfficeVerifier.crcAuto);
			} else {
				verifier.setRevocationCheck(OfficeVerifier.crcNone);
			}

			if (optext(args, "-ignoreerrors")) {
				verifier.setIgnoreChainValidationErrors(true);
			} else {
				verifier.setIgnoreChainValidationErrors(false);
			}

			if (optext(args, "-offline")) {
				verifier.setOfflineMode(true);
			} else {
				verifier.setOfflineMode(false);
			}

			verifier.verify();

			System.out.println("There are " + verifier.getSignatures().size() + " signatures in this file.");
			for (int x = 0; x < verifier.getSignatures().size(); x++) {
				System.out.println("Signature #" + (x + 1));
				System.out.println("  Signature type: " + translateSigType(verifier.getSignatures().item(x).getSignatureType()));
				System.out.println("  " + translateDocSig(verifier.getSignatures().item(x).getDocumentSigned()));
				System.out.println("  " + translateCore(verifier.getSignatures().item(x).getCorePropertiesSigned()));
				System.out.println("  " + translateOrigSig(verifier.getSignatures().item(x).getSignatureOriginSigned()));
				System.out.println("  Signature Time: " + verifier.getSignatures().item(x).getClaimedSigningTime());
				System.out.println("  Signature Validation Result: " + translateSigValidity(verifier.getSignatures().item(x).getSignatureValidationResult()));
				System.out.println("  Chain Validation Result: " + translateChainValidity(verifier.getSignatures().item(x).getChainValidationResult()) + "\n");
			}

			confirmExit();
		} catch (Exception ex) {
			displayError(ex);
		}
	}
}

class ConsoleDemo {
  private static BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));

  static String input() {
    try {
      return bf.readLine();
    } catch (IOException ioe) {
      return "";
    }
  }
  static char read() {
    return input().charAt(0);
  }

  static String prompt(String label) {
    return prompt(label, ":");
  }
  static String prompt(String label, String punctuation) {
    System.out.print(label + punctuation + " ");
    return input();
  }
  static String prompt(String label, String punctuation, String defaultVal) {
      System.out.print(label + " [" + defaultVal + "]" + punctuation + " ");
      String response = input();
      if (response.equals(""))
        return defaultVal;
      else
        return response;
  }

  static char ask(String label) {
    return ask(label, "?");
  }
  static char ask(String label, String punctuation) {
    return ask(label, punctuation, "(y/n)");
  }
  static char ask(String label, String punctuation, String answers) {
    System.out.print(label + punctuation + " " + answers + " ");
    return Character.toLowerCase(read());
  }

  static void displayError(Exception e) {
    System.out.print("Error");
    if (e instanceof SecureBlackboxException) {
      System.out.print(" (" + ((SecureBlackboxException) e).getCode() + ")");
    }
    System.out.println(": " + e.getMessage());
    e.printStackTrace();
  }

  /**
   * Takes a list of switch arguments or name-value arguments and turns it into a map.
   */
  static java.util.Map<String, String> parseArgs(String[] args) {
    java.util.Map<String, String> map = new java.util.HashMap<String, String>();
    
    for (int i = 0; i < args.length; i++) {
      // Add a key to the map for each argument.
      if (args[i].startsWith("-")) {
        // If the next argument does NOT start with a "-" then it is a value.
        if (i + 1 < args.length && !args[i + 1].startsWith("-")) {
          // Save the value and skip the next entry in the list of arguments.
          map.put(args[i].toLowerCase().replaceFirst("^-+", ""), args[i + 1]);
          i++;
        } else {
          // If the next argument starts with a "-", then we assume the current one is a switch.
          map.put(args[i].toLowerCase().replaceFirst("^-+", ""), "");
        }
      } else {
        // If the argument does not start with a "-", store the argument based on the index.
        map.put(Integer.toString(i), args[i].toLowerCase());
      }
    }
    return map;
  }
}

File: SecureBlackbox Samples/Office Verifier/c++/officeverifier.cpp
/*
 * SecureBlackbox 2024 C++ Edition - Sample Project
 *
 * This sample project demonstrates the usage of SecureBlackbox in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/secureblackbox
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "../../include/secureblackbox.h"

namespace ArgParser {
    static char* optval(int argc, char** argv, const char* option) {
        for (int x = 0; x < argc - 1; x++) {
            if (!strcmp(argv[x], option)) {
                return argv[x + 1];
            }
        }
        return (char*)"";
    }

    static bool optext(int argc, char** argv, const char* option) {
        for (int x = 0; x < argc; x++) {
            if (!strcmp(argv[x], option)) {
                return true;
            }
        }
        return false;
    }
};

using namespace ArgParser;

void displayHelp(const char* errMes) {
    printf(
        "NAME\n"
        "  officeverifier -- SecureBlackbox OfficeVerifier Demo Application\n\n"
        "SYNOPSIS\n"
		"  officeverifier <-input input_file> [-cert certificate_file] [-certpass certificate_password] [-checkrev] [-ignoreerrors] [-offline]\n"
        "DESCRIPTION\n"
        "  OfficeVerifier demonstrates the usage of OfficeVerifier from SecureBlackbox.\n"
        "  Used to verify the signature.\n\n"
        "  The options are as follows:\n\n"
        "  -input        An input file to verify (Required).\n\n"
        "  -cert         The certificate used to sign files.\n\n"
        "  -certpass     The password for the signing certificate.\n\n"
		"  -checkrev      Whether certificate revocation information should be checked.\n\n"
		"  -ignoreerrors  Whether to ignore chain validation errors.\n\n"
		"  -offline       Whether offline mode be used.\n\n"
		"EXAMPLES\n"
		"  officeverifier -input C:\\myfile.doc -output C:\\sign.doc -offline\n\n"
		"  officeverifier -input C:\\myfile.doc -output C:\\sign.doc -checkrev -ignoreerrors\n\n"
    );

	if (strlen(errMes) > 0)
	{
		printf("Error: %s \n\n", errMes);
	}
}

const char* translateSigType(int value) {
    switch (value) {
    case OST_BINARY_CRYPTO_API: return "BinaryCryptoAPI";
        break;
    case OST_BINARY_XML: return "BinaryXML";
        break;
    case OST_OPEN_XML: return "OpenXML";
        break;
    case OST_OPEN_XPS: return "OpenXPS";
        break;
    case OST_OPEN_DOCUMENT: return "OpenOffice";
        break;
    default: return "Unknown";
        break;
    }
}

const char* translateDocSig(bool value) {
    return value ? "Document content is signed" : "Document content is partially signed";
}

const char* translateCore(bool value) {
    return value ? "Document properties are signed" : "Document properties are not signed";
}

const char* translateOrigSig(bool value) {
    return value ? "Signature origin is signed" : "Signature origin is not signed";
}

const char* translateSigValidity(int value) {
    switch (value) {
    case SVT_VALID: return "Valid";
        break;
    case SVT_CORRUPTED: return "Corrupted";
        break;
    case SVT_SIGNER_NOT_FOUND: return "Signer not found";
        break;
    case SVT_FAILURE: return "Failure";
        break;
    default: return "Unknown";
        break;
    }
}

const char* translateChainValidity(int value) {
    switch (value) {
    case CVT_VALID: return "Valid";
        break;
    case CVT_VALID_BUT_UNTRUSTED: return "ValidButUntrusted";
        break;
    case CVT_INVALID: return "Invalid";
        break;
    case CVT_CANT_BE_ESTABLISHED: return "CantBeEstablished";
        break;
    default: return "Unknown";
        break;
    }
}

int main(int argc, char** argv) {
    OfficeVerifier verifier;
    CertificateManager cm;

    // Validate input
    if (argc < 2) {
        displayHelp("");
        goto done;
    }

    char* input = optval(argc, argv, "-input");
    if (!strcmp(input, "")) {
        displayHelp("-input is required.");
        goto done;
    }

    // Additional options
    char* cert = optval(argc, argv, "-cert");
    char* certpass = optval(argc, argv, "-certpass");

    // Required options
    verifier.SetInputFile(input);

    if (strcmp(cert, "")) {
        cm.ImportFromFile(cert, certpass);
        verifier.SetKnownCertCount(1);
        verifier.SetKnownCertHandle(0, cm.GetCertHandle());
    }

	optext(argc, argv, "-checkrev")
		? verifier.SetRevocationCheck(CRC_AUTO)
		: verifier.SetRevocationCheck(CRC_NONE);
	optext(argc, argv, "-ignoreerrors")
		? verifier.SetIgnoreChainValidationErrors(true)
		: verifier.SetIgnoreChainValidationErrors(false);
	optext(argc, argv, "-offline")
		? verifier.SetOfflineMode(true)
		: verifier.SetOfflineMode(false);

    // Verify
    if (verifier.Verify()) {
        goto err;
    }

    printf("There are %i signatures in this file.\n", verifier.GetSignatureCount());
    for (int x = 0; x < verifier.GetSignatureCount(); x++) {
        printf(
            "Signature #%i\n"
            "  Signature type: %s\n"
            "  %s\n"
            "  %s\n"
            "  %s\n"
            "  Signature Time: %s\n"
            "  Signature Validation Result: %s\n"
            "  Chain Validation Result: %s\n\n",
            x, translateSigType(verifier.GetSignatureSignatureType(x)),
            translateDocSig(verifier.GetSignatureDocumentSigned(x)),
            translateCore(verifier.GetSignatureCorePropertiesSigned(x)),
            translateOrigSig(verifier.GetSignatureSignatureOriginSigned(x)),
            verifier.GetSignatureClaimedSigningTime(x), translateSigValidity(verifier.GetSignatureSignatureValidationResult(x)),
            translateChainValidity(verifier.GetSignatureChainValidationResult(x))
        );
    }

err:
    if (verifier.GetLastErrorCode()) {
        printf("Error: [%i] %s\n", verifier.GetLastErrorCode(), verifier.GetLastError());
    }
done:
	printf("Press Enter to exit the demo.\n");
    getchar();
}


